docs/proposals/deployment.md

```diff
@@ -0,0 +1,182 @@
+[here](http://releases.k8s.io/release-1.0/docs/proposals/deployment.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Deployment
+
+## Abstract
+
+A proposal for implementing a new resource - Deployment - which will enable
+declarative config updates for Pods and ReplicationControllers.
+
+Users will be able to create a Deployment, which will spin of
+a ReplicationController to bring up the desired pods.
+Users can also target the Deployment at existing ReplicationControllers, in
+which case the new RC will replace the existing ones. The exact mechanics of
+replacement depends on the DeploymentStrategy chosen by the user.
+DeploymentStrategies are explained in detail in a later section.
+
+## Implementation
+
+### API Object
+
+The `Deployment` API object will have the following structure:
+
+```go
+type Deployment struct {
+  TypeMeta
+  ObjectMeta
+
+  // Specification of the desired behavior of the Deployment.
+  Spec DeploymentSpec
+
+  // Most recently observed status of the Deployment.
+  Status DeploymentStatus
+}
+
+type DeploymentSpec struct {
+  // Number of desired pods. This is a pointer to distinguish between explicit
+  // zero and not specified. Defaults to 1.
+  Replicas *int
+
+  // Label selector for pods. Existing ReplicationControllers which have all these
+  // label selectors will be scaled down.
+  Selector map[string]string
+
+  // Describes the pods that will be created.
+  Template *PodTemplateSpec
+
+  // The deployment strategy to use to replace existing pods with new ones.
+  Strategy DeploymentStrategy
+}
+
+type DeploymentStrategy struct {
+  // Type of deployment. Can be "Recreate" or "RollingUpdate".
+  Type DeploymentType
+
+  // Rolling update config params. Present only if DeploymentType =
+  // RollingUpdate.
+  RollingUpdate *RollingUpdateDeploymentSpec
+}
+
+type DeploymentType string
+
+const (
+  // Kill all existing pods before creating new ones.
+  DeploymentRecreate DeploymentType = "Recreate"
+
+  // Replace the old RCs by new one using rolling update i.e gradually scale down the old RCs and scale up the new one.
+  DeploymentRollingUpdate DeploymentType = "RollingUpdate"
+)
+
+// Spec to control the desired behavior of rolling update.
+type RollingUpdateDeploymentSpec struct {
+  // The maximum number of pods that can be unavailable during the update.
+  // Value can be an absolute number (ex: 5) or a percentage of total pods at the start of update (ex: 10%)
+  // By default, a fixed value of 1 is used.
+  maxUnavailable IntOrString
+
+  // The maximum number of pods that be scheduled over the original number of
+  // pods.
+  // Value can be an absolute number (ex: 5) or a percentage of total pods at
+  // the start of the update (ex: 10%)
+  // By default, a value of 1 is used.
+  maxSurge IntOrString
+
```

Comment:
cc @ironcladlou 
Dan's PR does not have this MinReadySeconds (minimum number of seconds for which the pod should be ready before it can be considered "truly ready").
Having LastTransitionTime in PodCondition will help in supporting this: https://github.com/kubernetes/kubernetes/issues/12615. Else, we will need to check that pod is ready, then wait for MinReadySeconds and then check the pod again to verify that it is still ready.

If we want the clock to be reset on cluster restart: then it seems like we need another condition something like "Stable". It will turn false as soon as a container crashes. It will turn back to true, if all containers come up. We can use the LastTransitionTime for Stable condition to know if the pod has been stable for the last x seconds.

I am wondering if all these conditions are required for RC as well? 
Do people want to create an RC to bring up X replicas and want RC to report a pod as up only if the pod has been ready for some minimum number of seconds. If RC has all that logic, then DC can just set RC.spec.replicas to the desired number of replicas and wait for RC.status.replicas to be set to that number. Right now, DC just ignores RC.status.